modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,872 @@
+    */
+    pragma "no doc"
+    proc _rotate(ref node: nodeType, pos: int) {
+      var child = node!.children[pos];
+      var parent = node!.parent;
+
+      _link(child!.children[pos^1], node, pos);
+      _link(node, child, pos^1);
+      
+
+      // Update the size field
+      node!.update();
+      child!.update();
+
+      child!.parent = parent;
+      node = child; // This is for change the node in its parent's children array
+    }
+
+    /*
+      Helper procedure to locate a certain node
+    */
+    pragma "no doc"
+    proc const _find(const node: nodeType, element: eltType): nodeType
+    lifetime return node {
+      if node == nil then return node;
+      var cmp = chpl_compare(element, node!.element, comparator);
+      if cmp == 0 then return node;
+      else if cmp < 0 then return _find(node!.children[0], element);
+      else return _find(node!.children[1], element);
+    }
+
+    /*
+      Helper procedure to locate a certain node
+      Returns a ref to the node
+    */
+    pragma "no doc"
+    proc _findRef(ref node: nodeType, element: eltType) ref: nodeType
+    lifetime return node {
+      if node == nil then return node;
+      var cmp = chpl_compare(element, node!.element, comparator);
+      if cmp == 0 then return node;
+      else if cmp < 0 then return _findRef(node!.children[0], element);
+      else return _findRef(node!.children[1], element);
+    }
+
+    /* Given one element, return the reference to the element in the orderedSet, 
+       which equals to the former in the perspective of the comparator.
+
+       This procedure could halt when there is no hit
+
+       Used by orderedMap
+     */
+    pragma "no doc"
+    proc _getReference(element: eltType) ref {
+      var node = _findRef(_root, element);
+      if node == nil then
+        boundsCheckHalt("index " + element:string + " out of bounds");
+      ref result = node!.element;
+      return result;
+    }
+
+    /* Given one element, return the element in the orderedSet, which equals to the 
+       former in the perspective of the comparator.
+
+       This procedure could halt when there is no hit
+
+       Used by orderedMap
+     */
+    pragma "no doc"
+    proc const _getValue(element: eltType) const {
+      var node = _find(_root, element);
+      if node == nil then
+        boundsCheckHalt("index " + element:string + " out of bounds");
+      var result = node!.element;
+      return result;
+    }
+
+    /*
+      Compare wrapper
+    */
+    pragma "no doc"
+    proc const _compare(x: eltType, y: eltType) {
+      return chpl_compare(x, y, comparator);
+    }
+
+    pragma "no doc"
+    proc ref _insert(ref node: nodeType, element: eltType, parent: nodeType): bool {
+      if node == nil {
+        node = new nodeType(element, _random(), 1, parent);
+        return true;
+      }
+      var cmp = _compare(element, node!.element);
+      if cmp == 0 then return false;
```

Comment:
Treap is a submodule of OrderedSet and designed to be like this. If there would be an OrderedMultiSet in the future and we want to reuse some part of OrderedSet, it will involve non-trival hierarchy changes like moving Treap out of OrderedSet. I think the refactoring should be done at that time.